### DISCUSSION TITLE
WARN: Use matchDepNames instead of matchPackageNames (yet again, maybe last time?)

### DISCUSSION THREAD

**Discussion by TWiStErRob** [renovatebot/renovate#29260] [Request Help]
### What would you like help with?

Other

### How are you running Renovate?

Mend Renovate hosted app on github.com

### Please tell us more about your question or problem

Would it be possible to put these suggestions in the warning message itself? When I read the logs all the information is available as a big json blob, but that's a non-trivial leap and still needs a lot of research on Renovate docs / discussions / issues.

The improved warning message could be:

> WARN: Use matchDepNames on `org.gradle.toolchains.foojay-resolver-convention`, or matchPackageNames on `org.gradle.toolchains.foojay-resolver-convention:org.gradle.toolchains.foojay-resolver-convention.gradle.plugin`; instead of matchPackageNames on `org.gradle.toolchains.foojay-resolver-convention`.

![image](https://github.com/renovatebot/renovate/assets/2906988/a6178348-dcfb-4d83-b8d5-bd8e32f158ac)
![image](https://github.com/renovatebot/renovate/assets/2906988/37a1b966-9924-473c-bbea-b98832ac5347)

I know it might be long, but at least it's easy to see what's going on, and the Dependency Dashboards / Renovate Dashboards / Logs would show actionable steps **directly**. It would also list **all** the problems at once, rather than having to play whack-a-mole (because of [`logger.once.warn`](https://github.com/renovatebot/renovate/blob/e521f7f3c030568d9cad991e532224081c0d2de0/lib/util/package-rules/package-names.ts#L24)'s `once` part.).

This would've prevented:
 * https://github.com/renovatebot/renovate/discussions/28753
 * https://github.com/renovatebot/renovate/discussions/25233
 * https://github.com/renovatebot/renovate/discussions/24623
 * https://github.com/renovatebot/renovate/discussions/23178

### Logs (if relevant)

```json
WARN: Use matchDepNames instead of matchPackageNames
{
  "packageRule": {
    "description": [
      "Auto-merge all updates of specific dependencies, their 0.x versions are stable and well-controlled."
    ],
    "matchUpdateTypes": [
      "minor",
      "patch"
    ],
    "matchCurrentVersion": "/^0/",
    "matchPackageNames": [
      "org.gradle.toolchains.foojay-resolver-convention"
    ],
    "automerge": true
  }
  "packageName": "org.gradle.toolchains.foojay-resolver-convention:org.gradle.toolchains.foojay-resolver-convention.gradle.plugin"
  "depName": "org.gradle.toolchains.foojay-resolver-convention"
}
```

**Comment by rarkins** [comment]
We have a policy of avoiding string interpolation in warning or error messages, so that backend log systems which count/aggregate such messages don't end up with infinite possibilities

**Reply by TWiStErRob** [reply]
Oh, that's unfortunate. Any chance of removing the `once`? So at least the message shows up multiple times in the logs?

I guess this would mean the dashboards also showing multiple (which could be a good thing), the deduplication probably should be at the UI level anyways, so that logs contain all the details, while the UIs show that "there is some problem(s)".

Alternatively keep the warn once, but also add debug without once?

**Reply by TWiStErRob** [reply]
Does the policy also apply to debug logs? That could be a nice way of providing all details while keeping the warning clean.

**Reply by rarkins** [reply]
Regarding "once", it was meant to include the metadata when it decides whether to log or suppress. If you have _different_ metadata being suppressed, then we should fix that so that each unique combination is logged.

Debug logs can definitely have string interpolation

**Reply by TWiStErRob** [reply]
Ok, will create a repro then to double-check.

**Reply by TWiStErRob** [reply]
Repro in https://github.com/TWiStErRob/renovate-29260

(Will delete it in a few weeks, in that case the source code will be archived at https://github.com/TWiStErRob/repros/tree/main/renovate/packageRules-matchDepNames-vs-matchPackageNames)

**Reply by rarkins** [reply]
Forked to https://github.com/renovate-reproductions/29260

@zharinov can you refresh my memory about logger.once and whether it could be adapted to incorporate metadata too?
